### ISSUE TITLE
Inconsistent colons?

### ISSUE THREAD

**Issue by StoneyJackson** [ourPLCC/plcc#37]
OK here is a question I have each year, but then I forget about it because it's not that big of a deal...

```
<tree>:Interior ::= LPAREN <SYMBOL> <tree>left <tree>right RPAREN
```

Why do we use a colon on the LHS (`<tree>:Interior`) and not on the RHS (`<tree>left`) when disambiguating terms?

Can we, should we, use the same syntax for both?

**Comment by jashelio** [comment]
In case you don’t realize, the notions are really not the same, although both have to do with disambiguation.

On the left, it is a totally new rule and the colon replaces the CFG notion of alternation. That is, in our system, it is written as a subclass, instead of this way:

Tree ::= Interior | Leaf | Root

On the right side, it provides unique names to the children of a “node”. This is not necessary for just a grammar, but it is in order to build the tree for semantic analysis.

Again, both features deal with code generation for semantic analysis (in different ways), but the latter feature has nothing to do with the grammar, whereas the former one does.


But after all that, I should say that I have no fundamental objection to making the two consistent in syntax. I don’t THINK it would confuse the students. That is, however, a code-breaking change, and therefore could be a real headache.

Jim

**Comment by fosler** [comment]
All,

One way to "fix" this is to allow for ":" to be optional in both cases. This would not break existing code, but it would permit

     <tree>Interior ::= LPAREN <SYMBOL> <tree>left <tree>right RPAREN

as well as

     <tree>:Interior ::= LPAREN <SYMBOL> <tree>:left <tree>:right RPAREN

On the other hand, Jim has some good points about justifying the current way. IMHO, we shouldn't change it. It's a PLCC "feature", not a bug.

Tim

**Comment by StoneyJackson** [comment]
Correct me if I'm wrong, but I don't think either of you are completely against the idea of having a consistent syntax for passing names to the code generator.

I think the main objections I'm hearing have to do with the disruption that the change might cause for instructors and their courses. I totally agree. And even if we never implement this specific change, I think we need to come up with a general policy/plan/strategy for rolling out breaking changes that minimizes disruption and helps instructors manage the rollout of breaking changes to their courses. Hopefully this would allow us to accept breaking changes because their minor benefit is not outweighed by their disruptive cost (note, we'd still have to weigh its implementation and maintenance cost!). When I get some more time I'll probably start a discussion on an issue. If you've got ideas about that and have time before me feel free to beat me to it :)

Closing this issue for now.

**Comment by fosler** [comment]
All,

I have tested a modified version of plcc.py that allows for an OPTIONAL ":" for name annotations in RHS rules. For example, using this version with the 'grammar' file in V6, you can say

     <exp>:IfExp    ::= IF <exp>:testExp THEN <exp>:trueExp ELSE <exp>:falseExp

Since the ":" in the RHS of such rules is optional, the rule

     <exp>:IfExp    ::= IF <exp>testExp THEN <exp>trueExp ELSE <exp>falseExp

still works. The colon in the LHS is still required.

This change adds TWO CHARACTERS to the 'defangg' procedure in plcc.py. [The documentation for defangg needs to be changed as well.]

If anyone thinks that this change is NEEDED or even HIGHLY DESIRABLE, or even OK, I will make the change to my plcc.py source in plcc.pithon.net, and Stoney can do his magic to incorporate this into a "feature' release in ourPLCC.

I am strongly disinclined to suggest that the ":" be optional for LHS entries, simply because I couldn't see how to make the change easily!

Let me know what you want to do...

Tim

**Comment by jashelio** [comment]
Sorry for my late reply. I still have reservations about this, but without strong authoritative arguments. If it feels good to you two, it's fine.

Jim

**Comment by StoneyJackson** [comment]
I want to make sure that we don't disrupt anyone's classes. I propose the following...

We create a `dev` branch separate from `main/master`. In that branch, we add new featured, breaking changes (bugs fixes are a little tricker, I'll come back to this). Then sometime during summer (early?), we could cut the release by merging `dev` into `main/master`.

It's important to create separate commits for bug fixes so they can be cherry picked between `main/master` and `dev`. If they are committed in the same commit as a feature change, then we would need to reproduce the fix manually on the different branches. Of course, there may be times when fixing a bug must be done differently in each branch because of changes in `dev`.

So, this new feature, the optional colon, would currently live only in the `dev` branch. All new development should only occur on `dev`. Only bug-fixes should be allowed to be applied to `main/master`.

@fosler if you plan to make the change in pithon.net and then I grab it and put it into the new `dev` branch, then your code in pithon.net is essential the `dev` branch in ourplcc/plcc.

(reopening this issue)

**Comment by StoneyJackson** [comment]
Scratch that... we should let `main/master` be the development branch and create release branches. Otherwise we'll be fighting our tools.

So `main/master` is for development. When we cut a new release, we'll create a branch for the release. We should limit ourselves to only adding non-breaking changes to a release branch. We should probably further restrict that to only bug fixes. That way releases are as stable as possible for instructors. In this model, when a bug is found, it's important to first check if the bug exists in a release, and make the fix to the release first. Then either merge it or cherry-pick the fix to main; if that's not possible, then make a custom fix for main.

I think I'll need to adjust the release mechanism and I might need to adjust the CI/CD configuration.

How does this sound?

**Comment by jashelio** [comment]
I am feeling that we need to meet for me to understand this process. Any interest?  - Jim

**Comment by StoneyJackson** [comment]
Sure. I'm wide open on Thursday.

**Comment by StoneyJackson** [comment]
Sure. I'm wide open on Thursday.

**Comment by StoneyJackson** [comment]
In anticipation of our meet up, I thought I would write a more expansive overview of what I'm suggesting and why we need it. For the images to render, you may need to view this on GitHub (https://github.com/ourPLCC/plcc/issues/37).

# Current

Below is a rough approximation of our current commit history. Circles are commits, ovals are tags, and rectangles are branches. So we have one branch, `master`. We have several tags, one for each release. In this diagram, time flows up.

> **Note:** GitHub makes it easy to download a zip for each tag (for example see <https://github.com/ourPLCC/plcc/tags>). So tagged commits are easy to distribute to non-Git users. Just what you want for each release.

![branch-model](https://plantuml.com/plantuml/svg/LP1B3i8m34JtFeLth5erIJyiK5N15RX0r4O5YRHI4ZW_qLY8PlLYUJBsU8h9XlJsBphnRVV-cIATkQf67cW0U-h1ZZgmfHOqP41D6bIam8QorTskaMBQWrqTVrJTr4rPgdzfQdNiwbAiIx66-FYvK8kN36Gi9TCOYuetOtanPWGILqP8SMG45pl6CNAeQx3R_72HPXiMzZOc5_AnBjNlFRbjoQVy0W00 "branch-model")

`master` points to a commit that is not in a release. It's version, stored in src/VERSION, is `3.1.1-dev.0`. That indicates it is a development copy and may contain changes that have not yet been officially released. When we decide to make a release, we tag `master` with a relesae tag.

> **More details on how a release is created**
>
> To create a release, we create a commit on `master` that has an updated CHANGELOG and an official release number in VERSION, and tag that commit.
>
> Right after re release, we create another commit on `master` that contains a development version in VERSION, making it safe to commit new changes to `master`.


# So far, so good

Each summer, new changes are proposed and added to `master`, and we cut releases as we go. By the end of summer we have a release that we are happy with, and that's the release for the year.

We've never had a severe bug that needed to be fixed immediately. But if we did, the process would be easy. Make new commits to `master` and cut a new patch release. Life is good.

# Something wicked this way comes

Now a potentially disruptive change has been proposed. We don't want to release it immeidately because instructors need time to prepare. So what do we do? Easy, add it to `master` and don't cut a release. Tell instructors about the coming change and that they can try it out by cloning our project and checking out `master`.

![branch-model-1](https://plantuml.com/plantuml/svg/JO_F2i8m38VlUOgmYwUEYRW_83BqxiaN26tP1hQL9jlp6pi4B_aI-C8l6LaeonfUSC1x2c4HXW6DkM8FVM47NM4BRU4PL7qayGnHArPkuRo-PTaysfdYv2iaHWSErHw-zfCc1fQKQVAudUfZtVoFZOuxiEJi1fPitRvbYdRME4lPOI0Md_N6Y_dreuzEF_W0 "branch-model-1")

But now, what happens when a bug is found in our currently supported release? If we fix it in master and cut a new release, that release also has the new disruptive change. The diagram below shows the problem (note, the new release is v4.0.0 because it contains a breaking change as well as the bugfix).

![branch-model-2](https://plantuml.com/plantuml/svg/JL1D2y8m3BtdLmGldZOwr7s0oD2x9_z0jjQkO3jfk-7FjsOJTidZ9UIzbxGXae-JVS4URwEr9WOe16RP1Hje66ke6Ike64ie6K-GTkvZL05SghZh9lqqdnrIm6vjIHFyz8vcLjWFvBJYgOGEaq0XG28R9Q47wG0XZfwqmlcG5xdOKh6bHwPB5pdHie2SSXcl1advwidrGqhPavTeAKJbVpUUipzfbPFf1Ly0 "branch-model-2")

The only way for an end user to get the bugfix is to upgrade to the new disruptive release.

# Release branches for the win!

To avoid this, we should fix the bug directly on the release that has the bug. We'll need a place in Git to do this. So we create a release branch. Suppose v3.1.0 contains the bug. Then we create a branch named v3.1 that initially points to the same commit that the v3.1.0 tag points.

![branch-model-3](https://plantuml.com/plantuml/svg/JO_D2eD038JlUOhWfIV5KkeFb2BjlQU-G70N5RghRABFtnWfz9APWIzCqh5GaZLyu8JFEOH965hCypisq9ZMK9jMK9bUGT7NB9uXwiJCJPpMHQRDOpzI77o6nEZ0eTBjJhzfO62P4mqUjqjnBih_M6ey1Atd8D2wTVkHAFQZrlMKdAqVBX2BJthjBV_vUlT6T3uw_UeB "branch-model-3")

We checkout v3.1 and commit a fix. Then we cut a new release v3.1.1.

![branch-model-4](https://plantuml.com/plantuml/svg/LL3B3i8W4BplLsnwyTIcnDX7OaoZTq_-mBOW9H5gW3P-lXHgr0ipiyjiPhTpdgoVzGDsU9cqLjv1Ytb-mWQQY3MmSYDrn0ggY0S8dklaXGCJNiouSdP-UhK874OoKcH83ldMxMTvLwzOwe5Z6D12lm8hrq4taWwSdon9WSk-O4NvAyj_oGe6gOenKJ9WJ9dQtybdIsOOGyg1B8_cn3GvBspOzfX_UDWqYFYb4uQ7ixm1 "branch-model-4")

One last problem. The fix now exists for `v3.1`, but not for `master` and the future release. So we need to either cherry-pick (a git operation) commit 9 to `master`, or merge `v3.1` into `master`. Let's to the latter.

![branch-model-5](https://plantuml.com/plantuml/svg/LL3B3eCW4BplLsoyUD90clf8ciQqzvxw0oWKJGesWAQVNrsqQI_Cp2uxEz3u85oOp1DIl8x63C43vvXbP-HiPpNKX1LKX2Lmjf6Ii822y0YBpMqCoeDTJapauDtq2iEiiEk5rIf1uL5ktNRIZ-5DfHOaqgOMu-hLwIwq1nz69xJ2-P3pdFrAzYzvpY5MaIB50QIOiRrRftXnmdRz4hCJJj9mP4RueDpwxbFs5MmLTATHLYv_zG40 "branch-model-5")

So now commit 10 contains the new change in 8 and the bug fix in 9, but 9 only contains the bug fix.

# What would we do without Git?

Git is allowing us to manage multiple "live" lines of development. This is difficult to do without Git.

**Comment by StoneyJackson** [comment]
Added in #38
